### ISSUE TITLE
[JENKINS-55512] Safe shutdown/restart should not block completion of complex jobs (that spawn child jobs)

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-23#8871]
<p>Jenkins supports "safe" behavior to shut down (safeExit) or restart (safeRestart) itself, which seems to be used in plugin updater, thinBackup plugin (doing just a quietDown part of this), operations that can be requested via Jenkins URL (and a link from Jenkins Manage interface), to name a few use-cases.</p>

<p>This mode waits for currently running jobs to complete and disallows new jobs to progress from scheduled to running. The assumption is that the current jobs will complete, the server will be quiet and can be administratively restarted with no severe interruption to its users. Documented in more detail at <a href="https://support.cloudbees.com/hc/en-us/articles/216118748-How-to-Start-Stop-or-Restart-your-Instance-" class="external-link" target="_blank" rel="nofollow noopener">https://support.cloudbees.com/hc/en-us/articles/216118748-How-to-Start-Stop-or-Restart-your-Instance-</a> for example.</p>

<p>This is problematic however when there are complex jobs, such as MultiPhase or entangled pipelines, where one wrapper job calls as its payload over time a number of other jobs that implement certain tests or other operations. When the safe shutdown mode is enabled, these child jobs can not be started, and the parent job stalls indefinitely waiting for their result, and the Jenkins master is left dysfunctional (not restarted for e.g. upgrade overnight, and not running any new builds).</p>

<p>The expected operational result would be that the currently running wrapper jobs AND any children (and their children) that can get spawned would be allowed to start and awaited to complete (boils down to "completion of all running jobs" as before), after which the safe restart/shutdown normally takes place.</p>

<p>I believe this could be done with some simple check of the build cause in the code which disallows execution of scheduled builds when the safe shutdown is enabled, to allow building of jobs triggered by a job, but would disallow builds triggered by SCM changes, Polling, Indexing, manually triggered (or maybe that one with an option to go through nonetheless?) etc. This seems like the place (or good starting point): <a href="https://github.com/jenkinsci/jenkins/blob/d5eeefd7beb8a00910f8d579ef14124a8be1914c/core/src/main/java/hudson/model/Queue.java#L1786" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/blob/d5eeefd7beb8a00910f8d579ef14124a8be1914c/core/src/main/java/hudson/model/Queue.java#L1786</a></p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jimklimov.png" title="jimklimov's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jimklimov">jimklimov</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-55512" target="_blank">Safe shutdown/restart should not block completion of complex jobs (that spawn child jobs)</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jimklimov.png" title="jimklimov's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jimklimov">jimklimov</a>
<li><b>status</b>: Open
<li><b>priority</b>: Major
<li><b>component(s)</b>: core
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 1
<li><b>watchers</b>: 6
<li><b>imported</b>: 2025-11-23
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Jenkins supports "safe" behavior to shut down (safeExit) or restart (safeRestart) itself, which seems to be used in plugin updater, thinBackup plugin (doing just a quietDown part of this), operations that can be requested via Jenkins URL (and a link from Jenkins Manage interface), to name a few use-cases.</p>

<p>This mode waits for currently running jobs to complete and disallows new jobs to progress from scheduled to running. The assumption is that the current jobs will complete, the server will be quiet and can be administratively restarted with no severe interruption to its users. Documented in more detail at <a href="https://support.cloudbees.com/hc/en-us/articles/216118748-How-to-Start-Stop-or-Restart-your-Instance-" class="external-link" target="_blank" rel="nofollow noopener">https://support.cloudbees.com/hc/en-us/articles/216118748-How-to-Start-Stop-or-Restart-your-Instance-</a> for example.</p>

<p>This is problematic however when there are complex jobs, such as MultiPhase or entangled pipelines, where one wrapper job calls as its payload over time a number of other jobs that implement certain tests or other operations. When the safe shutdown mode is enabled, these child jobs can not be started, and the parent job stalls indefinitely waiting for their result, and the Jenkins master is left dysfunctional (not restarted for e.g. upgrade overnight, and not running any new builds).</p>

<p>The expected operational result would be that the currently running wrapper jobs AND any children (and their children) that can get spawned would be allowed to start and awaited to complete (boils down to "completion of all running jobs" as before), after which the safe restart/shutdown normally takes place.</p>

<p>I believe this could be done with some simple check of the build cause in the code which disallows execution of scheduled builds when the safe shutdown is enabled, to allow building of jobs triggered by a job, but would disallow builds triggered by SCM changes, Polling, Indexing, manually triggered (or maybe that one with an option to go through nonetheless?) etc. This seems like the place (or good starting point): <a href="https://github.com/jenkinsci/jenkins/blob/d5eeefd7beb8a00910f8d579ef14124a8be1914c/core/src/main/java/hudson/model/Queue.java#L1786" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/blob/d5eeefd7beb8a00910f8d579ef14124a8be1914c/core/src/main/java/hudson/model/Queue.java#L1786</a></p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>from ages ago up till current Jenkins 2,156</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jimklimov.png" title="jimklimov's avatar" /> jimklimov:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-55512?focusedId=357454&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-357454">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Proposed PR <a href="https://github.com/jenkinsci/jenkins/pull/3840" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/pull/3840</a> to address this.</p>
</pre>
</details>

</ul>
</details>

<p>Proposed PR <a href="https://github.com/jenkinsci/jenkins/pull/3840" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/pull/3840</a> to address this.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jimklimov.png" title="jimklimov's avatar" /> jimklimov:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-55512?focusedId=359624&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-359624">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Upon discussion with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=danielbeck" class="user-hover" rel="danielbeck">danielbeck</a> the initial approach I took was not fully proper, at least not for upstreaming the change, as it would unilaterally change behavior for all child jobs, and this might not be applicable for some pipelines that can be (or not be) restartable, and certainly not intended for triggering of downstream jobs (rather than sub-jobs in focus of this issue) that have the same UpstreamCause &#8211; such jobs are independent and can wait in the blocked queue until the Jenkins master is safely restarted. This would also not be easily extensible to add behaviors (e.g. the idea about sysadmin forcing some job even though the server is shutting down).</p>

<p>The better suggestion was to make use of Actions, which are an attachable piece of metadata, e.g. to inherit some new nonBlockingJobAction (name is arbitrary), consider its presence when making Queue.java decisions, and attach it to jobs scheduled in the plugins like the Freestyle, MultiPhase build and Pipeline (Workflow) build steps <b>in case</b> the current job would wait for completion of that new child job (async children can also wait until restart of the master, right?).</p>

<p>For additional features, such as an administrative enforcement to run some job even if the server is quieting down, an extension point should be defined in the jenkins-core and then new plugins can be made to take advantage of that based on their purpose, logic and configurability. Effectively, the code in Queue.java would ask all currently loaded implementations of the extension point whether this Queue.Item (and/or Queue.Task) CAN be executed despite the pending shutdown, aka "is not blocked by the pending shutdown", with probably some tri-state outcome (permitted, blocked, no_opinion), with the current implementation being the fallback for a definitive result unless there is some certain verdict earlier.</p>

<ul>
	<li><a href="https://jenkins.io/doc/developer/extensibility/" class="external-link" target="_blank" rel="nofollow noopener">https://jenkins.io/doc/developer/extensibility/</a></li>
	<li><a href="https://wiki.jenkins-ci.org/display/JENKINS/AnnotationExtension" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.jenkins-ci.org/display/JENKINS/AnnotationExtension</a></li>
	<li><a href="https://wiki.jenkins-ci.org/display/JENKINS/Defining+a+new+extension+point" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.jenkins-ci.org/display/JENKINS/Defining+a+new+extension+point</a></li>
</ul>


<p>PS: Hope I got and relayed it right <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>Upon discussion with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=danielbeck" class="user-hover" rel="danielbeck">danielbeck</a> the initial approach I took was not fully proper, at least not for upstreaming the change, as it would unilaterally change behavior for all child jobs, and this might not be applicable for some pipelines that can be (or not be) restartable, and certainly not intended for triggering of downstream jobs (rather than sub-jobs in focus of this issue) that have the same UpstreamCause &#8211; such jobs are independent and can wait in the blocked queue until the Jenkins master is safely restarted. This would also not be easily extensible to add behaviors (e.g. the idea about sysadmin forcing some job even though the server is shutting down).</p>

<p>The better suggestion was to make use of Actions, which are an attachable piece of metadata, e.g. to inherit some new nonBlockingJobAction (name is arbitrary), consider its presence when making Queue.java decisions, and attach it to jobs scheduled in the plugins like the Freestyle, MultiPhase build and Pipeline (Workflow) build steps <b>in case</b> the current job would wait for completion of that new child job (async children can also wait until restart of the master, right?).</p>

<p>For additional features, such as an administrative enforcement to run some job even if the server is quieting down, an extension point should be defined in the jenkins-core and then new plugins can be made to take advantage of that based on their purpose, logic and configurability. Effectively, the code in Queue.java would ask all currently loaded implementations of the extension point whether this Queue.Item (and/or Queue.Task) CAN be executed despite the pending shutdown, aka "is not blocked by the pending shutdown", with probably some tri-state outcome (permitted, blocked, no_opinion), with the current implementation being the fallback for a definitive result unless there is some certain verdict earlier.</p>

<ul>
	<li><a href="https://jenkins.io/doc/developer/extensibility/" class="external-link" target="_blank" rel="nofollow noopener">https://jenkins.io/doc/developer/extensibility/</a></li>
	<li><a href="https://wiki.jenkins-ci.org/display/JENKINS/AnnotationExtension" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.jenkins-ci.org/display/JENKINS/AnnotationExtension</a></li>
	<li><a href="https://wiki.jenkins-ci.org/display/JENKINS/Defining+a+new+extension+point" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.jenkins-ci.org/display/JENKINS/Defining+a+new+extension+point</a></li>
</ul>


<p>PS: Hope I got and relayed it right <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-55512?focusedId=359637&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-359637">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Looks about right.</p>

<p>The main problem seems to be that <tt>isBlockedByShutdown</tt> takes a <tt>Task</tt> argument (corresponds to <tt>Job</tt>), so this would need to be deprecated and replaced in favor of something taking a <tt>Queue.Item</tt> (precursor to <tt>Run</tt>/<tt>Build</tt>). This way, queue items/builds of a job can have different shutdown blocking characteristics.</p>

<p>Outline of plan:</p>
<ol>
	<li>Replace <tt>isBlockedByShutdown</tt> with this new implementation.</li>
	<li>Create <tt>NonBlockedQueueItemAction</tt> and check for its presence in the new <tt>isBlockedByShutdown</tt></li>
	<li>Actually add these new actions to queue items where it makes sense:
	<ol>
		<li>pipeline-build-step plugin or parameterized-trigger could add it to <tt>scheduleBuild2</tt> if they wait for the result to prevent deadlocking.</li>
		<li>Alternatively, any plugin could contribute the action via <tt>Queue.QueueDecisionHandler</tt> and allow manual admin overrides and similar.</li>
	</ol>
	</li>
</ol>

</pre>
</details>

</ul>
</details>

<p>Looks about right.</p>

<p>The main problem seems to be that <tt>isBlockedByShutdown</tt> takes a <tt>Task</tt> argument (corresponds to <tt>Job</tt>), so this would need to be deprecated and replaced in favor of something taking a <tt>Queue.Item</tt> (precursor to <tt>Run</tt>/<tt>Build</tt>). This way, queue items/builds of a job can have different shutdown blocking characteristics.</p>

<p>Outline of plan:</p>
<ol>
	<li>Replace <tt>isBlockedByShutdown</tt> with this new implementation.</li>
	<li>Create <tt>NonBlockedQueueItemAction</tt> and check for its presence in the new <tt>isBlockedByShutdown</tt></li>
	<li>Actually add these new actions to queue items where it makes sense:
	<ol>
		<li>pipeline-build-step plugin or parameterized-trigger could add it to <tt>scheduleBuild2</tt> if they wait for the result to prevent deadlocking.</li>
		<li>Alternatively, any plugin could contribute the action via <tt>Queue.QueueDecisionHandler</tt> and allow manual admin overrides and similar.</li>
	</ol>
	</li>
</ol>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-55512?focusedId=359638&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-359638">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The other alternative was to add a new extension point specifically for this (perhaps similar to <tt>QueueTaskDispatcher</tt>?), and have it be implemented in plugins. The problem there might be that we need tri-state Allow/Deny/Don't Care for proper 'fallthrough' to the default implementation. Or perhaps not, unsure. This is a different approach form my previous comment's, and is what Jim's last paragraph partially refers to.</p>
</pre>
</details>

</ul>
</details>

<p>The other alternative was to add a new extension point specifically for this (perhaps similar to <tt>QueueTaskDispatcher</tt>?), and have it be implemented in plugins. The problem there might be that we need tri-state Allow/Deny/Don't Care for proper 'fallthrough' to the default implementation. Or perhaps not, unsure. This is a different approach form my previous comment's, and is what Jim's last paragraph partially refers to.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jimklimov.png" title="jimklimov's avatar" /> jimklimov:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-55512?focusedId=379445&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-379445">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>See also <a href="https://github.com/jenkinsci/jenkins/pull/3920" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/pull/3920</a></p>
</pre>
</details>

</ul>
</details>

<p>See also <a href="https://github.com/jenkinsci/jenkins/pull/3920" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/pull/3920</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jimklimov.png" title="jimklimov's avatar" /> jimklimov:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-55512?focusedId=379446&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-379446">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>For some more alternative approaches, as I recently dug in some other plugins, I wondered if a solution could come from a cannibalized (and adapted by a non-Java developer) copy of `lockable-resources-plugin`?</p>

<p>It already has ways to block startup of jobs (waiting for "this resource" which has no instances available) after all, so it seems feasible to just replace the logic from locked resources processing by what I originally proposed for jobs/builds/causes relationships, and allow or block the builds. In fact, it would be less work to make this behavior optional and used only by specific jobs rather than server-wide, because the lockable resource abilities are something one configures in a job (and it would be some effort to find and rip out the relevant bits, going this way).</p>
</pre>
</details>

</ul>
</details>

<p>For some more alternative approaches, as I recently dug in some other plugins, I wondered if a solution could come from a cannibalized (and adapted by a non-Java developer) copy of `lockable-resources-plugin`?</p>

<p>It already has ways to block startup of jobs (waiting for "this resource" which has no instances available) after all, so it seems feasible to just replace the logic from locked resources processing by what I originally proposed for jobs/builds/causes relationships, and allow or block the builds. In fact, it would be less work to make this behavior optional and used only by specific jobs rather than server-wide, because the lockable resource abilities are something one configures in a job (and it would be some effort to find and rip out the relevant bits, going this way).</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jimklimov.png" title="jimklimov's avatar" /> jimklimov:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-55512?focusedId=420601&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-420601">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Currently not a task "in progress" - I had a shot at in-core PR that did not survive discussions, and an idea still in the back of my mind about plugin to add such behavior. But not realized, so anyone willing to pick it up - is welcome to <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>Currently not a task "in progress" - I had a shot at in-core PR that did not survive discussions, and an idea still in the back of my mind about plugin to add such behavior. But not realized, so anyone willing to pick it up - is welcome to <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
